Implement import and export

thomaskrause opened the pull request:
<!-- Please refer to our contributing documentation for any questions on submitting a pull request, or let us know here if you need any help. -->

## Pull request type

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. 
Choose one of the following types (you can copy and paste them if you like)

- Bugfix
- Feature
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no API changes)
- Build related changes
- Documentation content changes
- Other (please describe it)

--> 

Feature

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: fixes #111, fixes #114, fixes #115 


## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->

The following formats can be imported and exported using the Pepper library and a simple graphical wizard
- EXMARaLDA 
- Paula XML

The Pepper framework is added as a regular dependency, but the plugins are JAR-files loaded and activated by the new bundle.
This allows us to use a fixed version and don't depend on the Pepper update mechanism.

## Does this introduce a breaking change?

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

N/A

## Other information

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->


## Pull request checklist

Please check if your PR fulfills the following requirements:
- [X] Tests for the changes have been added **or** the PR is neither a bug fix nor a feature
- [x] Docs have been reviewed and added / updated if needed **or** the PR is neither a bug fix nor a feature
- [x] The *Unreleased* section in CHANGELOG.md has been amended to reflect the changes in this PR if needed
- [x] Build (`mvn verify`) was run locally and any changes were pushed
- [x] The pull request is against the correct branch (`master` for bug fixes, `develop` for new functionality)
- [x] [Dependencies and citation templates](https://github.com/hexatomic/hexatomic/tree/develop/releng/templates) have been updated where necessary **or** there are no new dependencies

---

# Checklist for maintainers

## Releases

- Do **NOT** push the green merge button on GitHub.  
Follow the latest version of the [developer/maintainer docs](https://hexatomic.github.io/hexatomic/dev/) for making releases.
- Check that license and citation information for dependencies are complete (each dependency has a folder in `/THIRD-PARTY/`).

sdruskat commented:
I'm wondering whether it'd make sense to include some minimal test data for the formats we support for import & export.

From a code perspective, this would be outside the scope, as it would mostly test code in the pepper-plugins.
From a functionality perspective, though, it'd make sense in integration tests to check whether the correct info has made it into the UI, e.g., the project manager, and possibly even one of the editors.

What do you think, @thomaskrause?

sdruskat commented on docs/user/src/usage/formats.md:
```diff
@@ -0,0 +1,44 @@
+
+## Export to a different format
+
+Exporting a corpus is similar to importing one.
```
Need to be clear with terminology: When we say `corpus` here, does it export the Salt project? Or one corpus?
Also attains to the next line: what needs to be selected in order to be able to export, and what impact does the selection have on the export?

thomaskrause replied:
It means the whole project, I will change that to "corpus project". It does however add a perceived inconsistency with the "import corpus" part, even if this is practically not an issue since most importers except the Salt one only import corpora and have no concept of project.

sdruskat replied:
Yes, I see the perceived inconsistency. Would still prefer consistency for what we work with in Hexatomic (Salt projects). Could possibly get around this by calling it "corpus data" or "linguistic data" in the import section, but don't rate this as important enough :).

sdruskat reviewed (changes requested):
Some minor suggestions on the one hand, on the other hand two more impactful points:

1. (Re-)export doesn't work (should be fixed in the scope of this PR):
    1. Open example corpus (Salt)
    1. Export to PAULA
    1. Import from PAULA (this is where the gif starts)
    1. Export menu item is not active

See my comment on including test data, the pcc2 example provided by [corpus-tools.org](https://corpus-tools.org/corpora/pcc2_PAULA.zip) is small enough (93kB) to be included and test at least one case IMHO, or even just one of the two documents?

![export](https://user-images.githubusercontent.com/3007126/100775670-de0ef300-3403-11eb-8440-81c8d2a71593.gif)

2. Import bound to fail should
    - either detect wrong format earlier (not sure if heuristics would support that, but ideally the format declares its own format, right?),
    - or come up with a better explanation for failing imports.

This could be fixed in a future PR, and keep this one implementing the basics.

Example: Import a Salt directory (e.g., example-corpus) as PAULA. In this case, the log messages aren't clear ("No files are given", but would expect "No PAULA files found", or even "Files are not in the right format" or something).

![import](https://user-images.githubusercontent.com/3007126/100776407-ce43de80-3404-11eb-8d76-2f067c9d583c.gif)

sdruskat commented:
Also just tried importing the EXMARaLDA test corpus (<https://doi.org10.25592/uhhfdm.8364>), which failed for 77 documents (perhaps that's all of them, I haven't counted). Trying to import different sub-folders also fails.

This is technically in the scope of the EXMARaLDA modules, but we should think about whether it is perhaps a requirement that importing the test corpus works?

thomaskrause commented:
>     1. (Re-)export doesn't work (should be fixed in the scope of this PR):
>        
>        1. Open example corpus (Salt)
>        2. Export to PAULA
>        3. Import from PAULA (this is where the gif starts)
>        4. Export menu item is not active

This is actually intended and documented behavior right now. If you import a corpus, it will not have a location on disk (only a temporary one which will vanish when Hexatomic is closed), which is indicated by the "unsaved" marker. The export only works if the project is saved to a non-temporary disk (see "A project must be saved to the disk and must not have any unsaved changes before you can export it." in the documentation). We must tell Pepper from where to convert the files from, hence this restriction. This is also true for unsaved projects, even if they have a location: when the changes are not saved yet, Pepper has no way of converting the actual non-saved data. The current solution is to disable the export menu entry, but we could also enable it and ask the user to save the project files before she/he can actually perform the export.

For the other issue when not-autodetecting a format, this is actually already a strong indicator that something is wrong, since both formats currently have a quite good autodection. The user would also be free to manually select another format and click next, but I thought the conscious choice of the user is "warning" enough. If the user tells us that she/he is really sure this is this format, they should expect errors on the way. How to best report errors is a different topic though, since it is really difficult to extract this information from Pepper. 

We could distinguish the error message if no document was imported at all from the one were only some documents have an error and indicate to the user, that maybe the wrong format was selected. I will try to implement this "complete fail".

thomaskrause commented:
> Also just tried importing the EXMARaLDA test corpus (https://doi.org/10.25592/uhhfdm.8364), which failed for 77 documents (perhaps that's all of them, I haven't counted). Trying to import different sub-folders also fails.
> 
> This is technically in the scope of the EXMARaLDA modules, but we should think about whether it is perhaps a requirement that importing the ("official") test corpus works?
> 
> Options:
> 
>     * Adapt EXMARaLDA modules to work with the test corpus
> 
>     * Scope out better what "EXMARaLDA import" means for Hexatomic in this PR (i.e., which dataset does actually import successfully? Use that as test data and in the docs?)
> 
>     * Leave out EXMARaLDA import for now

Weird, I will check if these affects all files and is a systematic error (non-implemented feature of Exmaralda?) and then we can discuss the options.

thomaskrause commented:
> We could distinguish the error message if no document was imported at all from the one were only some documents have an error and indicate to the user, that maybe the wrong format was selected. I will try to implement this "complete fail".

This has been implemented in 340db9e2.

thomaskrause commented:
> > Also just tried importing the EXMARaLDA test corpus (https://doi.org/10.25592/uhhfdm.8364), which failed for 77 documents (perhaps that's all of them, I haven't counted). Trying to import different sub-folders also fails.
> > This is technically in the scope of the EXMARaLDA modules, but we should think about whether it is perhaps a requirement that importing the ("official") test corpus works?
> > Options:
> > ```
> > * Adapt EXMARaLDA modules to work with the test corpus
> > 
> > * Scope out better what "EXMARaLDA import" means for Hexatomic in this PR (i.e., which dataset does actually import successfully? Use that as test data and in the docs?)
> > 
> > * Leave out EXMARaLDA import for now
> > ```
> 
> Weird, I will check if these affects all files and is a systematic error (non-implemented feature of Exmaralda?) and then we can discuss the options.

Ok, so several things are happening here.
1. The corpus is in the "corpus manager format" of Exmaralda, with a lot of extra files in addition to `*.exb` files. These files and directories containing only filtered files should be ignored by the Exmaralda Importer or it should detect that this is not a "EXB" corpus
2. The actual `*.exb` files have generic meta-information with no name: `<ud-information attribute-name="" />`. Trying to add these empty named annotations to Salt fails with an exception. This must be fixed in the Exmaralda Importer to make it more robust.

sdruskat commented:
> The current solution is to disable the export menu entry, but we could also enable it and ask the user to save the project files before she/he can actually perform the export.

I would actually prefer that the menu item is active and the user is asked to save as you suggest.

> For the other issue when not-autodetecting a format, this is actually already a strong indicator that something is wrong, since both formats currently have a quite good autodection.

I see, but the user may not know this.

> The user would also be free to manually select another format and click next, but I thought the conscious choice of the user is "warning" enough. If the user tells us that she/he is really sure this is this format, they should expect errors on the way. How to best report errors is a different topic though, since it is really difficult to extract this information from Pepper.

What about informing users that auto-detection wasn't possible with a status update in the wizard? Is cheap enough and makes this clearer.

> We could distinguish the error message if no document was imported at all from the one were only some documents have an error and indicate to the user, that maybe the wrong format was selected. I will try to implement this "complete fail".

:+1:

sdruskat commented:
>     1. The corpus is in the "corpus manager format" of Exmaralda, with a lot of extra files in addition to `*.exb` files. These files and directories containing only filtered files should be ignored by the Exmaralda Importer or it should detect that this is not a "EXB" corpus. This confusion around what the "EXMARaLDA format" is was the reason I originally wrote something like "EXMARaLDA Partitur Editor format" in the documentation but then simplified it because I was afraid no one knows what I'm talking about (except maybe Thomas Schmidt). In addition to the "EXMARAaLDA Corpus Manager format" (Coma) there is also a specialized transcription format with the ending "exs".

I see. Yes, ignoring files or detecting as "non-EXB" corpus would be good to have in the import. Either way, I think the respective info should probably be patched through to the Hexatomic user via the wizard (e.g., let it say `some files in the import are ignored as they are not .exb files`, or `Corpus is not a pure EXB corpus` (former would be better I think, as it doesn't signal "I can' do this" but "I can do this, but I won't be able to do *everything*").

>     2. The actual `*.exb` files have generic meta-information with no name: `<ud-information attribute-name="" />`. Trying to add these empty named annotations to Salt fails with an exception. This must be fixed in the Exmaralda Importer to make it more robust.

:+1:

On the topic of whether the changes to the EXB Importer (shold it be renamed to `EXMARaLDA_EXB_Modules`?) should be in the scope of this PR: I don't see it. I'd rather approve the fact that import / export works with the discussed changes. The changes to the importer should be made and integrated asap, but that's a different scope.

What do you think?

thomaskrause commented:
> > The current solution is to disable the export menu entry, but we could also enable it and ask the user to save the project files before she/he can actually perform the export.
> 
> I would actually prefer that the menu item is active and the user is asked to save as you suggest.

Is implemented that way, was a little bit difficult to actually test it because it both needs mocked objects and the real SWT shell for the dialogs.

> 
> > For the other issue when not-autodetecting a format, this is actually already a strong indicator that something is wrong, since both formats currently have a quite good autodection.
> 
> I see, but the user may not know this.
> 
> > The user would also be free to manually select another format and click next, but I thought the conscious choice of the user is "warning" enough. If the user tells us that she/he is really sure this is this format, they should expect errors on the way. How to best report errors is a different topic though, since it is really difficult to extract this information from Pepper.
> 
> What about informing users that auto-detection wasn't possible with a status update in the wizard? Is cheap enough and makes this clearer.

There is now an error message indicating this.

sdruskat commented:
I'm assuming no, but is there a way to get rid of the lengthy error messages logged from the `PepperTest-Thread` (whose name feels a bit awkward to have here as well)?

```
22:13:47.894 [PepperTest-Thread] INFO org.corpus_tools.pepper.core.PepperOSGiRunner - ************************************************************************************************************************
*                                         ____                                                                         *
*                                        |  _ \ ___ _ __  _ __   ___ _ __                                              *
*                                        | |_) / _ \ '_ \| '_ \ / _ \ '__|                                             *
*                                        |  __/  __/ |_) | |_) |  __/ |                                                *
*                                        |_|   \___| .__/| .__/ \___|_|                                                *
*                                                  |_|   |_|                                                           *
*                                                                                                                      *
************************************************************************************************************************
* Pepper is a Salt model based converter for a variety of linguistic formats.                                          *
* For further information, visit: http://u.hu-berlin.de/saltnpepper                                                    *
* For contact write an eMail to:  saltnpepper@lists.hu-berlin.de                                                       *
************************************************************************************************************************


22:13:47.908 [PepperTest-Thread] INFO org.corpus_tools.pepper.core.PepperOSGiRunner - +----------------------------------------------------------------------------------------------------------------------+
|  configuration for Pepper                                                                                            |
+----------------------------------------------------------------------------------------------------------------------+
| pepper.detailedStatusReport             : true                                                                       |
| pepper.gcAfterDocumentSleep             : true                                                                       |
| pepper.memPolicy                        : MODERATE                                                                   |
| pepper.maxAmountOfProcessedSDocuments   : 10                                                                         |
+----------------------------------------------------------------------------------------------------------------------+

22:13:47.919 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'DoNothingImporter'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.919 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'DoNothingImporter'.
22:13:47.920 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'SaltXMLImporter'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.920 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'SaltXMLImporter'.
22:13:47.921 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'TextImporter'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.921 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'TextImporter'.
22:13:47.921 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'DoNothingImporter'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.921 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'DoNothingImporter'.
22:13:47.921 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'SaltXMLImporter'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.922 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'SaltXMLImporter'.
22:13:47.922 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'TextImporter'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.922 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'TextImporter'.
22:13:47.923 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'DOTManipulator'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.924 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'DOTManipulator'.
22:13:47.925 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'DoNothingManipulator'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.925 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'DoNothingManipulator'.
22:13:47.927 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'SaltValidator'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.927 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'SaltValidator'.
22:13:47.927 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'DOTManipulator'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.927 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'DOTManipulator'.
22:13:47.928 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'DoNothingManipulator'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.928 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'DoNothingManipulator'.
22:13:47.928 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'SaltValidator'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.928 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'SaltValidator'.
22:13:47.930 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'DOTExporter'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.930 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'DOTExporter'.
22:13:47.931 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'DoNothingExporter'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.931 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'DoNothingExporter'.
22:13:47.932 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'SaltXMLExporter'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.932 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'SaltXMLExporter'.
22:13:47.933 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'TextExporter'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.933 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'TextExporter'.
22:13:47.933 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'DOTExporter'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.933 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'DOTExporter'.
22:13:47.934 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'DoNothingExporter'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.934 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'DoNothingExporter'.
22:13:47.934 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'SaltXMLExporter'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.934 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'SaltXMLExporter'.
22:13:47.935 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot find location of resource folder for Pepper module 'TextExporter'. This could cause problems running the module. Tried to detect the location via system property 'osgi.bundles' which was 'reference:file:org.eclipse.osgi.compatibility.state_1.1.800.v20200511-1223.jar,reference:file:org.apache.felix.scr_2.1.16.v20200110-1820.jar@1:start,reference:file:org.eclipse.equinox.simpleconfigurator_1.3.500.v20200211-1505.jar@1:start'. 
22:13:47.935 [PepperTest-Thread] WARN org.corpus_tools.pepper.core.ModuleResolverImpl - Cannot set resource for Pepper module 'TextExporter'.
22:13:47.936 [PepperTest-Thread] INFO org.corpus_tools.pepper.core.PepperOSGiRunner - +-----+---------------------+----------------+------------+--------------------------------+---------------------------+
| no. | module-name         | module-version | module-type| formats                        | website                   |
+-----+---------------------+----------------+------------+--------------------------------+---------------------------+
| 1   | DoNothingImporter   | 3.5.3          | IMPORTER   | doNothing, 0.0                 | http://corpus-tools.      |
|     |                     |                |            |                                | org/pepper/               |
| 2   | SaltXMLImporter     | 3.5.3          | IMPORTER   | SaltXML, 1.0                   | http://corpus-tools.      |
|     |                     |                |            |                                | org/pepper/               |
| 3   | TextImporter        | 3.5.3          | IMPORTER   | txt, 0.0                       | http://corpus-tools.      |
|     |                     |                |            |                                | org/pepper/               |
| 4   | DOTManipulator      | 3.5.3          | MANIPULATOR|                                | http://corpus-tools.      |
|     |                     |                |            |                                | org/pepper/               |
| 5   | DoNothingManipulator| 3.5.3          | MANIPULATOR|                                | http://corpus-tools.      |
|     |                     |                |            |                                | org/pepper/               |
| 6   | SaltValidator       | 3.5.3          | MANIPULATOR|                                | http://corpus-tools.      |
|     |                     |                |            |                                | org/pepper/               |
| 7   | DOTExporter         | 3.5.3          | EXPORTER   | dot, 1.0                       | http://corpus-tools.      |
|     |                     |                |            |                                | org/pepper/               |
| 8   | DoNothingExporter   | 3.5.3          | EXPORTER   | doNothing, 0.0                 | http://corpus-tools.      |
|     |                     |                |            |                                | org/pepper/               |
| 9   | SaltXMLExporter     | 3.5.3          | EXPORTER   | SaltXML, 1.0                   | http://corpus-tools.      |
|     |                     |                |            |                                | org/pepper/               |
| 10  | TextExporter        | 3.5.3          | EXPORTER   | txt, 1.0                       | http://corpus-tools.      |
|     |                     |                |            |                                | org/pepper/               |
+-----+---------------------+----------------+------------+--------------------------------+---------------------------+

22:13:47.936 [PepperTest-Thread] ERROR org.corpus_tools.pepper.core.PepperOSGiRunner - An exception in Test occured. Cannot start pepper-osgi-runner, please set environment variable 'PEPPER_TEST_WORKFLOW_FILE' to workflow description file which is supposed to be used for conversion.
22:13:47.936 [PepperTest-Thread] INFO org.corpus_tools.pepper.core.PepperOSGiRunner - Conversion ended, and needed: 47 ms
22:13:47.936 [PepperTest-Thread] INFO org.corpus_tools.pepper.core.PepperOSGiRunner - ************************************************************************************************************************
22:13:48.417 [main] INFO  o.c.h.c.ApplicationLifecycle - Logging configured from logback.xml in the Hexatomic root folder
22:13:48.451 [main] INFO  o.c.h.f.Activator - Using /home/stephan/src/hexatomic/releng/org.corpus_tools.hexatomic.product/target/products/org.corpus_tools.hexatomic.product/linux/gtk/x86_64/configuration/org.eclipse.osgi/31/0/.cp/pepper-plugins as Pepper module resource folder
22:13:48.452 [main] INFO  o.c.h.f.Activator - Installing Pepper module from file:/home/stephan/src/hexatomic/releng/org.corpus_tools.hexatomic.product/target/products/org.corpus_tools.hexatomic.product/linux/gtk/x86_64/configuration/org.eclipse.osgi/31/0/.cp/pepper-plugins/exmaralda-emf-api-1.2.2.jar
22:13:48.474 [main] INFO  o.c.h.f.Activator - Installing Pepper module from file:/home/stephan/src/hexatomic/releng/org.corpus_tools.hexatomic.product/target/products/org.corpus_tools.hexatomic.product/linux/gtk/x86_64/configuration/org.eclipse.osgi/31/0/.cp/pepper-plugins/pepperModules-EXMARaLDAModules-1.3.1.jar
22:13:48.486 [main] INFO  o.c.h.f.Activator - Installing Pepper module from file:/home/stephan/src/hexatomic/releng/org.corpus_tools.hexatomic.product/target/products/org.corpus_tools.hexatomic.product/linux/gtk/x86_64/configuration/org.eclipse.osgi/31/0/.cp/pepper-plugins/pepperModules-PAULAModules-1.3.0.jar
```

sdruskat commented:
Suggest to remove the VSCode-specific files in `docs/.vscode/...`.

sdruskat commented on docs/user/src/usage/import-export.md:
```diff
@@ -0,0 +1,47 @@
+Currently, Hexatomic supports the following external file formats:
+
+- [PAULA XML](https://github.com/korpling/paula-xml)
+- [EXMARaLDA files](https://exmaralda.org)
```
```suggestion
- [EXMARaLDA *.exb files](https://exmaralda.org)
```

I think the current labeling on the import wizard is clear enough for users to realize they can only import .exb files, but this should be made clear here as well.

sdruskat commented on docs/user/src/usage/import-export.md:
```diff
@@ -0,0 +1,47 @@
+Exporting a corpus project is similar to importing one.
+You select *Export* from the *File* menu, select an output path and the output format and finish the wizard.
+The output path should not contain the root corpus name, since this will be appended automatically.
+A project must be saved to the disk and must not have any unsaved changes before you can export it.
```
```suggestion
A project must be saved to disk and must not have any unsaved changes before you can export it.
```

Remove unidiomatic `the`

sdruskat commented on docs/user/src/usage/import-export.md:
```diff
@@ -0,0 +1,47 @@
+
+![Import directory selection with the last directory name `example-corpus`](import-select-directory.png)
+
+Most formats use a single file for each document, but PAULA XML will use a sub-directory containing files for each annotation layer.
```
```suggestion
Most formats use a single file for each document, with the exception of PAULA XML, which uses a sub-directory containing files for each annotation layer.
```

Fix phrasing.

sdruskat commented on docs/user/src/usage/import-export.md:
```diff
@@ -0,0 +1,47 @@
+
+After clicking on *Finish* the import will start and show the progress.
+The wizard will close automatically after the import process finishes.
+You can also abort the import process by clicking on the red squared stop button right next to the progress bar.
```
```suggestion
You can also abort the import process by clicking on the red square *stop* button right next to the progress bar.
```

Fix wording

sdruskat commented on tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/formats/exporter/TestExportHandler.java:
```diff
@@ -0,0 +1,192 @@
+/**
+ * Tests the several conditions of the {@link ExportHandler}. The handler is used as a fixture with
+ * mocked dependencies instead of the real ones from the context to make it easier to access the
+ * state. We still need to do this tests in the integration test package, since we need access to
```
```suggestion
 * state. We still need to do these tests in the integration test package, since we need access to
```

Fix typo

sdruskat reviewed (approved):
W00T!

Nice one, thanks!

Just a few minor things, then please feel free to merge :tada:.

I think it is now sufficiently clear that we can only import .exb, and I like the other fixes for the last review comments.

thomaskrause commented:
> Suggest to remove the VSCode-specific files in `docs/.vscode/...`.

Removed and added an ignore entry, so it won't be added again. There are some nice plugins for spellchecking markdown files with <https://languagetool.org> in VS Code, that's why they were added the first place.

thomaskrause commented:
> I'm assuming no, but is there a way to get rid of the lengthy error messages logged from the `PepperTest-Thread` (whose name feels a bit awkward to have here as well)?


Nope, Pepper does not really care about any other environment it is running in. We could try to silence all messages, but this could also lead to problems when we have to debug real-life conversion problems. Proper error handling and better integration as a library are the main reasons why I prototyped a new implementation of the framework (<https://github.com/korpling/pepper-ng>). But this is future tech we can't rely on yet :)

sdruskat commented:
> > I'm assuming no, but is there a way to get rid of the lengthy error messages logged from the `PepperTest-Thread` (whose name feels a bit awkward to have here as well)?
> 
> Nope, Pepper does not really care about any other environment it is running in. We could try to silence all messages, but this could also lead to problems when we have to debug real-life conversion problems. Proper error handling and better integration as a library are the main reasons why I prototyped a new implementation of the framework (https://github.com/korpling/pepper-ng). But this is future tech we can't rely on yet :)

Would've thought so 😉. Guess people will learn to live with it, as we all have, haha.
